// @1inch/fusion-sdk/dist/fusion-order/settlement-post-interaction-data/settlement-post-interaction-data.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const limit_order_sdk_1 = require("@1inch/limit-order-sdk");
const settlement_post_interaction_data_1 = require("./settlement-post-interaction-data");
const utils_1 = require("../../sdk/utils");
describe('PostInteractionData', () => {
    it('Should encode/decode with no fees and whitelist', () => {
        const data = settlement_post_interaction_data_1.SettlementPostInteractionData.new({
            bankFee: 0n,
            auctionStartTime: 1708117482n,
            whitelist: []
        });
        expect(settlement_post_interaction_data_1.SettlementPostInteractionData.decode(data.encode())).toStrictEqual(data);
    });
    it('Should encode/decode with fees and whitelist', () => {
        const data = settlement_post_interaction_data_1.SettlementPostInteractionData.new({
            bankFee: 0n,
            auctionStartTime: 1708117482n,
            whitelist: [
                {
                    address: limit_order_sdk_1.Address.ZERO_ADDRESS,
                    delay: 0n
                }
            ],
            integratorFee: {
                receiver: limit_order_sdk_1.Address.ZERO_ADDRESS,
                ratio: (0, utils_1.bpsToRatioFormat)(10)
            }
        });
        expect(settlement_post_interaction_data_1.SettlementPostInteractionData.decode(data.encode())).toStrictEqual(data);
    });
});
