// @1inch/fusion-sdk/dist/utils/bytes/bytes-iter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytesIter = void 0;
const assert_1 = __importDefault(require("assert"));
const validations_1 = require("../../validations");
const utils_1 = require("../../utils");
class BytesIter {
    constructor(bytes) {
        (0, assert_1.default)((0, validations_1.isHexBytes)(bytes), 'invalid bytes value');
        this.bytes = bytes.slice(2);
    }
    isEmpty() {
        return this.bytes.length === 0;
    }
    nextByte() {
        return this.nextBytes(1);
    }
    nextBytes(n) {
        const cnt = n * 2;
        if (this.bytes.length < cnt) {
            throw new Error(`Can not consume ${n} bytes, have only ${this.bytes.length / 2}`);
        }
        const bytes = this.bytes.slice(0, cnt);
        this.bytes = this.bytes.slice(cnt);
        return BigInt((0, utils_1.add0x)(bytes));
    }
    nextUint8() {
        return this.nextByte();
    }
    nextUint16() {
        return this.nextBytes(2);
    }
    nextUint24() {
        return this.nextBytes(3);
    }
    nextUint32() {
        return this.nextBytes(4);
    }
    nextUint160() {
        return this.nextBytes(20);
    }
}
exports.BytesIter = BytesIter;
