// @1inch/fusion-sdk/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UINT_256_MAX = exports.UINT_24_MAX = exports.UINT_32_MAX = exports.UINT_40_MAX = exports.UINT_80_MAX = exports.UINT_160_MAX = exports.ONE_INCH_LIMIT_ORDER_V4 = exports.NetworkEnum = exports.ZX = void 0;
exports.ZX = '0x';
var NetworkEnum;
(function (NetworkEnum) {
    NetworkEnum[NetworkEnum["ETHEREUM"] = 1] = "ETHEREUM";
    NetworkEnum[NetworkEnum["POLYGON"] = 137] = "POLYGON";
    NetworkEnum[NetworkEnum["BINANCE"] = 56] = "BINANCE";
    NetworkEnum[NetworkEnum["ARBITRUM"] = 42161] = "ARBITRUM";
    NetworkEnum[NetworkEnum["AVALANCHE"] = 43114] = "AVALANCHE";
    NetworkEnum[NetworkEnum["OPTIMISM"] = 10] = "OPTIMISM";
    NetworkEnum[NetworkEnum["FANTOM"] = 250] = "FANTOM";
    NetworkEnum[NetworkEnum["GNOSIS"] = 100] = "GNOSIS";
    NetworkEnum[NetworkEnum["COINBASE"] = 8453] = "COINBASE";
})(NetworkEnum = exports.NetworkEnum || (exports.NetworkEnum = {}));
exports.ONE_INCH_LIMIT_ORDER_V4 = '0x111111125421ca6dc452d289314280a0f8842a65';
exports.UINT_160_MAX = (1n << 160n) - 1n;
exports.UINT_80_MAX = (1n << 80n) - 1n;
exports.UINT_40_MAX = (1n << 40n) - 1n;
exports.UINT_32_MAX = (1n << 32n) - 1n;
exports.UINT_24_MAX = (1n << 24n) - 1n;
exports.UINT_256_MAX = (1n << 256n) - 1n;
