// @1inch/fusion-sdk/dist/connector/blockchain/web3-provider.connector.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_mockito_1 = require("ts-mockito");
const limit_order_sdk_1 = require("@1inch/limit-order-sdk");
const web3_provider_connector_1 = require("./web3-provider-connector");
describe('Web3 provider connector', () => {
    let web3Provider;
    let web3ProviderConnector;
    const limitOrder = {
        salt: '618054093254',
        makerAsset: '0xe9e7cea3dedca5984780bafc599bd69add087d56',
        takerAsset: '0x111111111117dc0aa78b770fa6a738034120c302',
        maker: '0xfb3c7eb936cAA12B5A884d612393969A557d4307',
        receiver: '0x0000000000000000000000000000000000000000',
        makingAmount: '1000000000000000000',
        takingAmount: '1000000000000000000',
        makerTraits: '0'
    };
    const typedData = {
        primaryType: 'Order',
        types: {
            EIP712Domain: limit_order_sdk_1.EIP712Domain,
            Order: limit_order_sdk_1.Order
        },
        domain: {
            name: limit_order_sdk_1.LimitOrderV4TypeDataName,
            version: limit_order_sdk_1.LimitOrderV4TypeDataVersion,
            chainId: 1,
            verifyingContract: (0, limit_order_sdk_1.getLimitOrderContract)(1)
        },
        message: limitOrder
    };
    beforeEach(() => {
        web3Provider = (0, ts_mockito_1.mock)();
        web3ProviderConnector = new web3_provider_connector_1.Web3ProviderConnector((0, ts_mockito_1.instance)(web3Provider));
    });
    it('should call eth_signTypedData_v4 rpc method', async () => {
        const walletAddress = '0xasd';
        const extendedWeb3 = {
            signTypedDataV4: jest.fn()
        };
        (0, ts_mockito_1.when)(web3Provider.extend((0, ts_mockito_1.anything)())).thenReturn(extendedWeb3);
        await web3ProviderConnector.signTypedData(walletAddress, typedData);
        expect(extendedWeb3.signTypedDataV4).toHaveBeenCalledWith(walletAddress, JSON.stringify(typedData));
    });
});
