// @1inch/fusion-sdk/dist/auction-calculator/auction-calculator.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const limit_order_sdk_1 = require("@1inch/limit-order-sdk");
const auction_calculator_1 = require("./auction-calculator");
const fusion_order_1 = require("../fusion-order");
const sdk_1 = require("../sdk");
describe('Auction Calculator', () => {
    it('should be created successfully from suffix and salt', () => {
        const auctionStartTime = 1708448252n;
        const postInteraction = fusion_order_1.SettlementPostInteractionData.new({
            integratorFee: {
                ratio: (0, sdk_1.bpsToRatioFormat)(1),
                receiver: limit_order_sdk_1.Address.fromBigInt(1n)
            },
            bankFee: 0n,
            auctionStartTime,
            whitelist: []
        });
        const auctionDetails = new fusion_order_1.AuctionDetails({
            startTime: auctionStartTime,
            initialRateBump: 50000,
            duration: 120n,
            points: []
        });
        const calculator = auction_calculator_1.AuctionCalculator.fromAuctionData(postInteraction, auctionDetails);
        const rate = calculator.calcRateBump(Number(auctionStartTime + 60n));
        const auctionTakingAmount = calculator.calcAuctionTakingAmount('1420000000', rate);
        expect(rate).toBe(25000);
        expect(auctionTakingAmount).toBe('1423692355');
    });
});
