// @1inch/fusion-sdk/dist/sdk/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FusionSDK = void 0;
const limit_order_sdk_1 = require("@1inch/limit-order-sdk");
const encoders_1 = require("./encoders");
const api_1 = require("../api");
const orders_1 = require("../api/orders");
class FusionSDK {
    constructor(config) {
        this.config = config;
        this.api = api_1.FusionApi.new({
            url: config.url,
            network: config.network,
            httpProvider: config.httpProvider,
            authKey: config.authKey
        });
    }
    async getActiveOrders({ page, limit } = {}) {
        const request = orders_1.ActiveOrdersRequest.new({ page, limit });
        return this.api.getActiveOrders(request);
    }
    async getOrderStatus(orderHash) {
        const request = orders_1.OrderStatusRequest.new({ orderHash });
        return this.api.getOrderStatus(request);
    }
    async getOrdersByMaker({ limit, page, address }) {
        const request = orders_1.OrdersByMakerRequest.new({ limit, page, address });
        return this.api.getOrdersByMaker(request);
    }
    async getQuote(params) {
        const request = api_1.QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: limit_order_sdk_1.Address.ZERO_ADDRESS.toString(),
            permit: params.permit,
            enableEstimate: false,
            fee: params?.takingFeeBps,
            source: params.source
        });
        return this.api.getQuote(request);
    }
    async getQuoteWithCustomPreset(params, body) {
        const paramsRequest = api_1.QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: limit_order_sdk_1.Address.ZERO_ADDRESS.toString(),
            permit: params.permit,
            enableEstimate: false,
            fee: params?.takingFeeBps,
            source: params.source
        });
        const bodyRequest = api_1.QuoterCustomPresetRequest.new({
            customPreset: body.customPreset
        });
        return this.api.getQuoteWithCustomPreset(paramsRequest, bodyRequest);
    }
    async createOrder(params) {
        const quote = await this.getQuoteResult(params);
        if (!quote.quoteId) {
            throw new Error('quoter has not returned quoteId');
        }
        const order = quote.createFusionOrder({
            receiver: params.receiver
                ? new limit_order_sdk_1.Address(params.receiver)
                : undefined,
            preset: params.preset,
            nonce: params.nonce,
            permit: params.permit,
            takingFeeReceiver: params.fee?.takingFeeReceiver,
            allowPartialFills: params.allowPartialFills,
            allowMultipleFills: params.allowMultipleFills
        });
        const domain = (0, limit_order_sdk_1.getLimitOrderV4Domain)(this.config.network);
        const hash = order.getOrderHash(domain);
        return { order, hash, quoteId: quote.quoteId };
    }
    async submitOrder(order, quoteId) {
        if (!this.config.blockchainProvider) {
            throw new Error('blockchainProvider has not set to config');
        }
        const orderStruct = order.build();
        const domain = (0, limit_order_sdk_1.getLimitOrderV4Domain)(this.config.network);
        const signature = await this.config.blockchainProvider.signTypedData(orderStruct.maker, order.getTypedData(domain));
        const relayerRequest = api_1.RelayerRequest.new({
            order: orderStruct,
            signature,
            quoteId
        });
        await this.api.submitOrder(relayerRequest);
        return {
            order: orderStruct,
            signature,
            quoteId,
            orderHash: order.getOrderHash(domain)
        };
    }
    async placeOrder(params) {
        const { order, quoteId } = await this.createOrder(params);
        return this.submitOrder(order, quoteId);
    }
    async buildCancelOrderCallData(orderHash) {
        const getOrderRequest = orders_1.OrderStatusRequest.new({ orderHash });
        const orderData = await this.api.getOrderStatus(getOrderRequest);
        if (!orderData) {
            throw new Error(`Can not get order with the specified orderHash ${orderHash}`);
        }
        const { order } = orderData;
        return (0, encoders_1.encodeCancelOrder)(orderHash, new limit_order_sdk_1.MakerTraits(BigInt(order.makerTraits)));
    }
    async getQuoteResult(params) {
        const quoterRequest = api_1.QuoterRequest.new({
            fromTokenAddress: params.fromTokenAddress,
            toTokenAddress: params.toTokenAddress,
            amount: params.amount,
            walletAddress: params.walletAddress,
            permit: params.permit,
            enableEstimate: true,
            fee: params.fee?.takingFeeBps,
            source: params.source
        });
        if (!params.customPreset) {
            return this.api.getQuote(quoterRequest);
        }
        const quoterWithCustomPresetBodyRequest = api_1.QuoterCustomPresetRequest.new({
            customPreset: params.customPreset
        });
        return this.api.getQuoteWithCustomPreset(quoterRequest, quoterWithCustomPresetBodyRequest);
    }
}
exports.FusionSDK = FusionSDK;
